judge/views/api/api_v2.py

```diff
@@ -1,122 +1,613 @@
-            'rank': rank,
-            'rating': result.participation_rating,
-        })
-
-    resp['contests'] = {
-        "current_rating": last_rating[0].rating if last_rating else None,
-        "volatility": last_rating[0].volatility if last_rating else None,
-        'history': contest_history,
-    }
-
-    solved_problems = []
-    attempted_problems = []
-
-    problem_data = (Submission.objects.filter(points__gt=0, user=profile, problem__is_public=True,
-                                              problem__is_organization_private=False)
-                    .annotate(max_pts=Max('points'))
-                    .values_list('max_pts', 'problem__points', 'problem__code')
-                    .distinct())
-    for awarded_pts, max_pts, problem in problem_data:
-        if awarded_pts == max_pts:
-            solved_problems.append(problem)
+from django.conf import settings
+from django.core.exceptions import PermissionDenied, ValidationError
+from django.db.models import Count, F, OuterRef, Prefetch, Q, Subquery
+from django.http import Http404, JsonResponse
+from django.utils import timezone
+from django.utils.functional import cached_property
+from django.views.generic.detail import BaseDetailView
+from django.views.generic.list import BaseListView
+
+from judge.models import (
+    Contest, ContestParticipation, ContestTag, Organization, Problem, ProblemType, Profile, Rating, Submission,
+)
+from judge.utils.raw_sql import join_sql_subquery, use_straight_join
+from judge.views.submission import group_test_cases
+
+
+class APILoginRequiredException(Exception):
+    pass
+
+
+class APILoginRequiredMixin:
+    def setup_api(self, request, *args, **kwargs):
+        if not request.user.is_authenticated:
+            raise APILoginRequiredException()
+        super().setup_api(request, *args, **kwargs)
+
+
+class APIMixin:
+    @cached_property
+    def _now(self):
+        return timezone.now()
+
+    def get_object_data(self, obj):
+        raise NotImplementedError()
+
+    def get_api_data(self, context):
+        raise NotImplementedError()
+
+    def get_base_response(self, **kwargs):
+        resp = {
+            'api_version': '2.0',
+            'method': self.request.method.lower(),
+            'fetched': self._now.isoformat(),
+        }
+        resp.update(kwargs)
+        return resp
+
+    def get_data(self, context):
+        return self.get_base_response(data=self.get_api_data(context))
+
+    def get_error(self, exception):
+        caught_exceptions = {
+            ValueError: (400, 'invalid filter value type'),
+            ValidationError: (400, 'invalid filter value type'),
+            PermissionDenied: (403, 'permission denied'),
+            APILoginRequiredException: (403, 'login required'),
+            Http404: (404, 'page/object not found'),
+        }
+        exception_type = type(exception)
+        if exception_type in caught_exceptions:
+            exception_data = caught_exceptions[exception_type]
```

Comment:
Why not tuple unpack this?

Reply:
This has been updated in the latest commit.